### PR TITLE
base64 librariy for support mcuboot serial FWU 

### LINKED ISSUE #3081: Concise Binary Object Representation (CBOR)
**_Reported by Anas Nashif:_**

from RFC7049:

The Concise Binary Object Representation (CBOR) is a data format
whose design goals include the possibility of extremely small code
size, fairly small message size, and extensibility without the need
for version negotiation.  These design goals make it different from
earlier binary serializations such as ASN.1 and MessagePack.

(Imported from Jira ZEP-1620)

### PR DESCRIPTION
**Pull request by nvlsianpu**
This patch introduce libraries which are need for support Mcuboot serial recovery mode.
These libraries are in usage by newtmgr protocol encoding and parsing. Mcuboot serial recovery protocol simply implements image related piece of newtmgr protocol.

Libraries:

    base64: base64 library

These libraries will be useful for future OTA DFU works as well.

As the serial recovery functionality consumes additional flash, the bootloader partition was extended as well for nRF52840 SoC.

Signed-off-by: Andrzej Puzdrowski andrzej.puzdrowski@nordicsemi.no

### CODE REVIEW

File: ext/lib/base64/Kconfig [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Kconfig - tinycbor library support configuration options
```
Comment (mbolivar) [inline ext/lib/base64/Kconfig]: This looks like a copy/paste issue

File: ext/lib/base64/src/base64.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
+/*
+ * This file is based on roken from the FreeBSD source.  It has been modified
```
Comment (mbolivar) [inline ext/lib/base64/src/base64.c]: This is BSD licensed, not Apache

Comment (nvlsianpu) [reply ext/lib/base64/src/base64.c]: so I should change the commit message right?

File: ext/lib/base64/Kconfig (hunk 1 above)
Comment (nvlsianpu) [reply ext/lib/base64/Kconfig]: :D

File: ext/Kconfig [hunk 3]
```diff
@@ -14,4 +14,6 @@ source "ext/lib/crypto/Kconfig"
```
Comment (jukkar) [inline ext/Kconfig]: I wonder if we really need to have another base64 library as we have base64 support functions already in mbedtls. One is be able to use them without embedding full mbedtls library.

Comment (nvlsianpu) [reply ext/Kconfig]: Surely i was thinking about such option. What i think is that mbetls is another external library, not targeted to provide this encoding. So base64 from mbedtls might changes or disappears without giving us warning prior. (mbedtls is under arm subsidiary control) Is better to have base64 separately.

Comment (jukkar) [reply ext/Kconfig]: > So base64 from mbedtls might changes or disappears without giving us warning prior.

Not likely as there are already users of base64() in zephyr, and even if mbedtls would remove base64(), we could then add another library implementing that. Anyway, just my random comment to this while looking various PRs.

**Comment by pfalcon** [comment]
Ok for base64, but why we need hex encoding/decoding here?

**Comment by nvlsianpu** [comment]
As was agreed on discussion we have during the Open Source Summit Europe in Prague this patch is replaced by the import of mbedtls-base64 library. (https://github.com/zephyrproject-rtos/zephyr/pull/4743)

**Pull request closed without merging**